### PR TITLE
Reduce transpile time by 80% using TypeScript 3.4

### PR DESCRIPTION
**Pull request by cschroeter**
**Motivation for the change**
The Nest projects I'm working with takes about 18 seconds to re-transpile. (~ 10k lines of code).
With the changes in this PR I could manage to reduce transpile time to ~3 seconds. Even with this small starter project you'll notice the difference.

I have taken the liberty of also clean up some imperfections 😇

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,45 +4,43 @@
...
     "@nestjs/core": "6.0.2",
     "@nestjs/microservices": "6.0.2",
     "@nestjs/platform-express": "6.0.2",
     "@nestjs/websockets": "6.0.2",
-    "reflect-metadata": "0.1.13",
-    "rimraf": "2.6.3",
     "rxjs": "6.4.0"
   },
   "devDependencies": {
+    "@nestjs/testing": "6.0.1",
```
Comment (Sacro) [inline package.json]: Why is this out of order?

File: package.json [hunk 2]
```diff
@@ -4,45 +4,43 @@
...
+    "start:debug": "tsc-watch --onSuccess 'node --inspect-brk dist/main.js'",
+    "prestart:prod": "npm run build",
     "start:prod": "node dist/main.js",
-    "lint": "tslint -p tsconfig.json -c tslint.json",
     "test": "jest",
     "test:watch": "jest --watch",
     "test:cov": "jest --coverage",
-    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
-    "test:e2e": "jest --config ./test/jest-e2e.json"
+    "test:debug": "node --inspect-brk node_modules/.bin/jest --runInBand",
```
Comment (Sacro) [inline package.json]: Why have you removed  `-r tsconfig-paths/register -r ts-node/register` ?

File: package.json [hunk 3]
```diff
@@ -4,45 +4,43 @@
   "description": "Nest TypeScript starter repository",
   "license": "MIT",
   "scripts": {
-    "build": "tsc -p tsconfig.build.json",
-    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
```
Comment (Sacro) [inline package.json]: Why remove?

File: package.json [hunk 4]
```diff
@@ -4,45 +4,43 @@
   "description": "Nest TypeScript starter repository",
   "license": "MIT",
   "scripts": {
-    "build": "tsc -p tsconfig.build.json",
```
Comment (Sacro) [inline package.json]: Why remove?

**Comment by kamilmysliwiec** [comment]
Why `tsconfig.build` has been deleted? + 3.4.0 is in Release Candidate stage, not final one.

File: package.json (hunk 3 above)
Comment (cschroeter) [reply package.json]: `format` and `build` have not been removed. I just moved `format` and `lint` to the bottom of the scripts list and everyhting related to `start` the app to the top

File: package.json (hunk 1 above)
Comment (cschroeter) [reply package.json]: Did not touch that line, but can update to `6.0.2`

**Comment by cschroeter** [comment]
@kamilmysliwiec I am aware that TypeScript 3.4 has not been released - nevertheless the performance increate in terms of re-transpile time might be worth the update - or just keep this PR open until TS 3.4 has been released.

Regarding `tsconfig.build.json` the content of that file has been merged with `tsconfig.json`.
There is no need to compile tests while using the TypesScript compiler. The tests will still be transpield using `ts-node`

File: package.json (hunk 4 above)
Comment (cschroeter) [reply package.json]: The `build` script is still there. The content of `tsconfig.build.json` has been merged with `tsconfig.json`

File: package.json (hunk 2 above)
Comment (cschroeter) [reply package.json]: `ts-node` and `tsconfig-paths` have been replaced with `ts-watch`

**Comment by kamilmysliwiec** [comment]
> There is no need to compile tests while using the TypesScript compiler. The tests will still be transpield using ts-node

`tsconfig.build.json` has a reason for being there. These two shouldn't be merged.

**Comment by cschroeter** [comment]
> 
> `tsconfig.build.json` has a reason for being there. These two shouldn't be merged.

I am not aware why there is a need for a `tsconfig.json` and a `tsconfig.build.json`. The _new_ `build` script will return the same transpiled JavaScript then the `build` script before. Maybe you can explain what's the motivation here.

**Comment by kamilmysliwiec** [comment]
Isn't it quite dangerous to remove files if you don't know why they exist? ;) We need two separate `tsconfig` files to provide compatibility with IDEs that automatically use `tsconfig.json` file to provide, for instance, type-checking. In your PR, `tsconfig` will exclude `spec` files which means that some IDEs will no longer recognise TS in test files.

**Comment by cschroeter** [comment]
Shouldn't a starter project be IDE agnostic? TypeScript is configured with `tsconfig.json`. That should be it. The fact that there are two tsconfig files creates confusion and adds complexity that - imho - should be avoided.

**Comment by kamilmysliwiec** [comment]
@cschroeter Firstly, the starter project is intended to *work*.

**Comment by cschroeter** [comment]
OK I think we have different opinions about software quality.

**Comment by kamilmysliwiec** [comment]
Very likely :)

**Comment by cschroeter** [comment]
I care and u don't :)

**Comment by kamilmysliwiec** [comment]
😂

**Comment by cschroeter** [comment]
Just keep `rimraf` and `reflect-metadata` as regular dependencies. Nuff said 😂

**Pull request closed without merging**